### PR TITLE
Feature/sycl cell binning

### PR DESCRIPTION
**Pull request by will-saunders-ukaea**
# Description

Maps particles into Nektar++ mesh cells (2D linear) using SYCL.

- New feature (non-breaking change which adds functionality)

# Testing

- New test for mesh class.
- Passes existing tests with hipsycl/mpich + intel mpi//dpcpp.
- Passes tests (except 1D1V implementation which has compliance issues I think) with hipsycl + cuda-nvcxx.

### CODE REVIEW

**Review by oparry-ukaea — COMMENTED** [review]
Looks good to me.
The suggested changes are mostly typos, making some function args const, or doxygen pedantry. 

Unit and integration tests all pass for me with gcc and oneapi, although I haven't tried with a GPU backend.

**Comment by will-saunders-ukaea** [comment]
> > Passes tests (except 1D1V implementation which has compliance issues I think) with hipsycl + cuda-nvcxx.
> 
> Which test is that?

The TwoStream and FFT tests - Essentially any tests where host arrays are modified directly rather than using a host accessor (which is all of them). 

What NESO (the original 1D1V code) does is create std::vectors then create buffer member variables using the std::vector::data pointers (e.g. this one [1]). These buffers then have the same lifetime as the object. Tests modify data in the original vector directly not via a buffer host accessor. 

However the Buffer spec says "The ownership of this memory is given to the constructed SYCL buffer for the duration of its lifetime.". Hence when we modify the original vector directly this is non-compliant. The SYCL implementation doesn't see these writes and hence doesn't copy the new values to the device - hence the test fails when the buffer memory is separate from the host memory e.g. on the GPU.

[1] https://github.com/ExCALIBUR-NEPTUNE/NESO/blob/b8a75f7804c66494049614d195578428b6077d26/include/mesh.hpp#L54

**Comment by oparry-ukaea** [comment]
Got it, thanks for the explanation.

**Comment by jwscook** [comment]
> > > Passes tests (except 1D1V implementation which has compliance issues I think) with hipsycl + cuda-nvcxx.
> > 
> > 
> > Which test is that?
> 
> The TwoStream and FFT tests - Essentially any tests where host arrays are modified directly rather than using a host accessor (which is all of them).
> 
> What NESO (the original 1D1V code) does is create std::vectors then create buffer member variables using the std::vector::data pointers (e.g. this one [1]). These buffers then have the same lifetime as the object. Tests modify data in the original vector directly not via a buffer host accessor.
> 
> However the Buffer spec says "The ownership of this memory is given to the constructed SYCL buffer for the duration of its lifetime.". Hence when we modify the original vector directly this is non-compliant. The SYCL implementation doesn't see these writes and hence doesn't copy the new values to the device - hence the test fails when the buffer memory is separate from the host memory e.g. on the GPU.
> 
> [1]
> 
> https://github.com/ExCALIBUR-NEPTUNE/NESO/blob/b8a75f7804c66494049614d195578428b6077d26/include/mesh.hpp#L54

Is this encapsulated in an issue anywhere?

**Comment by will-saunders-ukaea** [comment]
It is now: https://github.com/ExCALIBUR-NEPTUNE/NESO/issues/171

File: test/unit/nektar_interface/test_utility_cartesian_mesh.cpp [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  }
+
+  std::array<double, 6> bb;
+  std::array<int, 2> cell = {1, 1};
+  mesh.get_bounding_box(cell, bb);
+
+  ASSERT_NEAR(bb[0], 0.0, 1.0e-10);
+  ASSERT_NEAR(bb[1], 0.0, 1.0e-10);
+  ASSERT_NEAR(bb[0 + 3], 1.0, 1.0e-10);
+  ASSERT_NEAR(bb[1 + 3], 2.0, 1.0e-10);
```
[the remark this answers is not in the document]
Comment (will-saunders-ukaea) [reply test/unit/nektar_interface/test_utility_cartesian_mesh.cpp:44]: Nothing deeper than pick a tolerance between 1e-8 and 1e-15.

[the remark this answers is not in the document]
Comment (will-saunders-ukaea) [reply test/unit/nektar_interface/test_utility_cartesian_mesh.cpp]: Made 1E-14 and ran formatting.

**Review by oparry-ukaea — APPROVED** [review]
Ready to merge.

**Pull request merged**